tests/test_url.py

```diff
@@ -1235,11 +1235,35 @@ def test_with_path_query():
     assert str(url.with_path("/test")) == "http://example.com/test"
 
 
+def test_with_path_keep_query():
+    url = URL("http://example.com?a=b")
+    url2 = url.with_path("/test", keep_query=True)
+    assert str(url2) == "http://example.com/test?a=b"
+
+
 def test_with_path_fragment():
     url = URL("http://example.com#frag")
     assert str(url.with_path("/test")) == "http://example.com/test"
 
 
+def test_with_path_keep_fragment():
+    url = URL("http://example.com#frag")
+    url2 = url.with_path("/test", keep_fragment=True)
+    assert str(url2) == "http://example.com/test#frag"
+
+
+def test_with_path_keep_fragment_and_query():
+    url = URL("http://example.com?a=b#frag")
+    url2 = url.with_path("/test", keep_query=True, keep_fragment=True)
+    assert str(url2) == "http://example.com/test?a=b#frag"
+
+
+def test_with_path_keep_fragment_and_query_false():
+    url = URL("http://example.com?a=b#frag")
+    url2 = url.with_path("/test", keep_query=False, keep_fragment=False)
+    assert str(url2) == "http://example.com/test"
```

Comment:
Looks like these could be collapsed via `@pytest.mark.parametrize()`. Plus, there's the case of `keep_query=False` with `keep_fragment=False` that seems to be missing, but could be added as another param.